src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,395 @@
+                    ProcessRemovals();
+                    _safeToDisposeHandleEvent.Set();
+
+                    // Indices may have changed when processing removals and the signalled handle may have already been unregistered
+                    // so we do a linear search over the active user waits to see if the signaled handle is still registered
+                    if (signaledHandleIndex != WaitHandle.WaitTimeout)
+                    {
+                        for (int i = 0; i < _numUserWaits; i++)
+                        {
+                            RegisteredWaitHandle registeredHandle = _registeredWaitHandles[i];
+                            if (registeredHandle.Handle == signaledHandle)
+                            {
+                                QueueWaitCompletion(registeredHandle, false);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        for (int i = 0; i < _numUserWaits; i++)
+                        {
+                            RegisteredWaitHandle registeredHandle = _registeredWaitHandles[i];
+                            if (registeredHandle.Timeout == _currentTimeout)
+                            {
+                                QueueWaitCompletion(registeredHandle, true);
+                            }
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Go through the <see cref="_pendingRemoves"/> array and remove those registered wait handles from the <see cref="_registeredWaitHandles"/>
+            /// and <see cref="_waitHandles"/> arrays. Then move elements around in those arrays to fill holes.
+            /// </summary>
+            private void ProcessRemovals()
+            {
+                _removesLock.Acquire();
+                _registeredHandlesLock.Acquire();
+                if(_numPendingRemoves == 0)
+                {
+                    _registeredHandlesLock.Release();
+                    _removesLock.Release();
+                    return;
+                }
+
+                // This is O(N^2), but max(N) = 63 and N will usually be very low
+                for (int i = 0; i < _numPendingRemoves; i++)
+                {
+                    for (int j = 0; j < _numUserWaits; j++)
+                    {
+                        if (_pendingRemoves[i] == _registeredWaitHandles[j])
+                        {
+                            _registeredWaitHandles[j] = null;
+                            _waitHandles[j + 1] = null;
+                            break;
+                        }
+                    }
+                    _pendingRemoves[i] = null;
+                }
+                _numPendingRemoves = 0;
+
+                // Fill in nulls
+                // This is O(1), Goes through each of the 63 possible handles once.
```

Comment:
N = _numUserWaits in the big-O notation, so it's constant time with respect to _numUserWaits.